If server-acl-init fails to pass out server tokens to every server it will not retry when re-run

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request. Searching for pre-existing feature requests helps us consolidate datapoints for identical requirements into a single place, thank you!
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

<!--- Thank you for keeping this note for the community --->

---

### Overview of the Issue

When `server-acl-init` runs it creates server ACL tokens and passes them out to servers (https://github.com/hashicorp/consul-k8s/blob/main/control-plane/subcommand/server-acl-init/servers.go#L113-L156). If this fails, e.g. one of the servers is stuck in legacy ACLs, then eventually the `server-acl-init` Job will timeout (5m). A new Pod will be created since that's the current Job configuration and that Pod will see that there is a bootstrap ACL token and so it will assume that all servers have been correctly bootstrapped and skip the server bootstrap step. 

The logs would look something like:

First run
```
2021-10-28T20:46:25.581Z [INFO]  Retrying in 1s
2021-10-28T20:46:26.610Z [INFO]  Success: bootstrapping ACLs - PUT /v1/acl/bootstrap
2021-10-28T20:46:26.624Z [INFO]  Success: writing bootstrap Secret "consul-bootstrap-acl-token"
2021-10-28T20:46:26.648Z [INFO]  Success: creating agent policy - PUT /v1/acl/policy
2021-10-28T20:46:26.659Z [INFO]  Success: creating server token for consul-server-0.consul-server.default.svc - PUT /v1/acl/token
2021-10-28T20:46:26.663Z [INFO]  Success: updating server token for consul-server-0.consul-server.default.svc - PUT /v1/agent/token/agent
2021-10-28T20:46:26.746Z [INFO]  Success: creating server token for consul-server-1.consul-server.default.svc - PUT /v1/acl/token
2021-10-28T20:46:26.749Z [INFO]  Success: updating server token for consul-server-1.consul-server.default.svc - PUT /v1/agent/token/agent
2021-10-28T20:46:26.843Z [ERROR] Failure: creating server token for consul-server-2.consul-server.default.svc - PUT /v1/acl/token: err="Unexpected response code: 500 (The ACL system is currently in legacy mode.)"
2021-10-28T20:46:26.843Z [INFO]  Retrying in 1s
2021-10-28T20:46:27.846Z [ERROR] Failure: creating server token for consul-server-2.consul-server.default.svc - PUT /v1/acl/token: err="Unexpected response code: 500 (The ACL system is currently in legacy mode.)"
2021-10-28T20:46:27.846Z [INFO]  Retrying in 1s
...
2021-10-28T20:53:54.841Z [ERROR] reached command timeout
```

Second run:

```
021-10-28T20:54:00.363Z [INFO]  ACLs already bootstrapped - retrieved bootstrap token from Secret "consul-bootstrap-acl-token"
2021-10-28T20:54:00.763Z [INFO]  Success: calling /agent/self to get datacenter
2021-10-28T20:54:00.764Z [INFO]  Current datacenter: datacenter=dc1 primaryDC=dc1
2021-10-28T20:54:00.766Z [INFO]  Policy "agent-token" already exists, updating
2021-10-28T20:54:00.774Z [INFO]  Success: creating agent policy - PUT /v1/acl/policy
2021-10-28T20:54:00.782Z [INFO]  Success: creating cross-namespace-policy policy
2021-10-28T20:54:00.869Z [INFO]  Success: creating client-token policy
2021-10-28T20:54:00.879Z [INFO]  Success: creating token for policy client-token
2021-10-28T20:54:00.887Z [INFO]  Success: writing Secret for token client-token
2021-10-28T20:54:00.894Z [INFO]  Success: creating anonymous token policy - PUT /v1/acl/policy
2021-10-28T20:54:00.903Z [INFO]  Success: updating anonymous token with policy
2021-10-28T20:54:00.963Z [INFO]  Success: getting consul-connect-injector-authmethod-svc-account ServiceAccount
2021-10-28T20:54:00.976Z [INFO]  Success: getting consul-connect-injector-authmethod-svc-account-token-abcd Secret
2021-10-28T20:54:00.987Z [INFO]  Success: creating auth method consul-k8s-auth-method
2021-10-28T20:54:00.989Z [INFO]  Success: listing binding rules for auth method consul-k8s-auth-method
2021-10-28T20:54:00.996Z [INFO]  Success: creating acl binding rule for consul-k8s-auth-method
2021-10-28T20:54:01.006Z [INFO]  Success: creating connect-inject-token policy
2021-10-28T20:54:01.069Z [INFO]  Success: creating token for policy connect-inject-token
2021-10-28T20:54:01.075Z [INFO]  Success: writing Secret for token connect-inject-token
2021-10-28T20:54:01.170Z [INFO]  Success: creating controller-token policy
2021-10-28T20:54:01.180Z [INFO]  Success: creating token for policy controller-token
2021-10-28T20:54:01.187Z [INFO]  Success: writing Secret for token controller-token
2021-10-28T20:54:01.187Z [INFO]  server-acl-init completed successfully
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,12 @@
 ## UNRELEASED
 
+BUG FIXES:
+* Control Plane
+  * ACLs: Fix issue where if one or more servers fail to have their ACL tokens set on the initial run of server-acl-init
+    then on subsequent re-runs of server-acl-init the tokens are never set. [[GH-825](https://github.com/hashicorp/consul-k8s/issues/825)]
+  * ACLs: Fix issue where if the number of Consul servers is increased, the new servers are never provisioned
+    an ACL token. [[GH-677](https://github.com/hashicorp/consul-k8s/issues/677)]
+
 ## 0.36.0 (November 02, 2021)
 
 BREAKING CHANGES:

--- a/control-plane/subcommand/server-acl-init/command.go
+++ b/control-plane/subcommand/server-acl-init/command.go
@@ -11,11 +11,6 @@ import (
 	"sync"
 	"time"
 
-	"github.com/hashicorp/consul-k8s/control-plane/consul"
-	"github.com/hashicorp/consul-k8s/control-plane/subcommand"
-	"github.com/hashicorp/consul-k8s/control-plane/subcommand/common"
-	"github.com/hashicorp/consul-k8s/control-plane/subcommand/flags"
-	k8sflags "github.com/hashicorp/consul-k8s/control-plane/subcommand/flags"
 	"github.com/hashicorp/consul/api"
 	"github.com/hashicorp/go-discover"
 	"github.com/hashicorp/go-hclog"
@@ -24,6 +19,12 @@ import (
 	k8serrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
+
+	"github.com/hashicorp/consul-k8s/control-plane/consul"
+	"github.com/hashicorp/consul-k8s/control-plane/subcommand"
+	"github.com/hashicorp/consul-k8s/control-plane/subcommand/common"
+	"github.com/hashicorp/consul-k8s/control-plane/subcommand/flags"
+	k8sflags "github.com/hashicorp/consul-k8s/control-plane/subcommand/flags"
 )
 
 type Command struct {
@@ -313,7 +314,6 @@ func (c *Command) Run(args []string) int {
 		scheme = "https"
 	}
 
-	var updateServerPolicy bool
 	var bootstrapToken string
 
 	if c.flagBootstrapTokenFile != "" {
@@ -340,19 +340,13 @@ func (c *Command) Run(args []string) int {
 
 		if bootstrapToken != "" {
 			c.log.Info(fmt.Sprintf("ACLs already bootstrapped - retrieved bootstrap token from Secret %q", bootTokenSecretName))
-
-			// Mark that we should update the server ACL policy in case
-			// there are namespace related config changes. Because of the
-			// organization of the server token creation code, the policy
-			// otherwise won't be updated.
-			updateServerPolicy = true
 		} else {
 			c.log.Info("No bootstrap token from previous installation found, continuing on to bootstrapping")
-			bootstrapToken, err = c.bootstrapServers(serverAddresses, bootTokenSecretName, scheme)
-			if err != nil {
-				c.log.Error(err.Error())
-				return 1
-			}
+		}
+		bootstrapToken, err = c.bootstrapServers(serverAddresses, bootstrapToken, bootTokenSecretName, scheme)
+		if err != nil {
+			c.log.Error(err.Error())
+			return 1
 		}
 	}
 
@@ -383,18 +377,6 @@ func (c *Command) Run(args []string) int {
 	}
 	c.log.Info("Current datacenter", "datacenter", consulDC, "primaryDC", primaryDC)
 	isPrimary := consulDC == primaryDC
-
-	// With the addition of namespaces, the ACL policies associated
-	// with the server tokens may need to be updated if Enterprise Consul
-	// users upgrade to 1.7+. This updates the policy if the bootstrap
-	// token had previously existed, which signals a potential config change.
-	if updateServerPolicy {
-		_, err = c.setServerPolicy(consulClient)
-		if err != nil {
-			c.log.Error("Error updating the server ACL policy", "err", err)
-			return 1
-		}
-	}
 
 	if c.flagEnablePartitions && c.flagPartitionName == consulDefaultPartition && isPrimary {
 		// Partition token is local because only the Primary datacenter can have Admin Partitions.

--- a/control-plane/subcommand/server-acl-init/command_test.go
+++ b/control-plane/subcommand/server-acl-init/command_test.go
@@ -15,10 +15,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/hashicorp/consul-k8s/control-plane/helper/cert"
-	"github.com/hashicorp/consul-k8s/control-plane/helper/go-discover/mocks"
-	"github.com/hashicorp/consul-k8s/control-plane/helper/test"
-	"github.com/hashicorp/consul-k8s/control-plane/subcommand/common"
 	"github.com/hashicorp/consul/api"
 	"github.com/hashicorp/consul/sdk/freeport"
 	"github.com/hashicorp/consul/sdk/testutil"
@@ -31,6 +27,11 @@ import (
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes/fake"
+
+	"github.com/hashicorp/consul-k8s/control-plane/helper/cert"
+	"github.com/hashicorp/consul-k8s/control-plane/helper/go-discover/mocks"
+	"github.com/hashicorp/consul-k8s/control-plane/helper/test"
+	"github.com/hashicorp/consul-k8s/control-plane/subcommand/common"
 )
 
 var ns = "default"
@@ -1284,6 +1285,8 @@ func TestRun_NoLeader(t *testing.T) {
 			numACLBootCalls++
 		case "/v1/agent/self":
 			fmt.Fprintln(w, `{"Config": {"Datacenter": "dc1", "PrimaryDatacenter": "dc1"}}`)
+		case "/v1/acl/tokens":
+			fmt.Fprintln(w, `[]`)
 		default:
 			fmt.Fprintln(w, "{}")
 		}
@@ -1341,6 +1344,10 @@ func TestRun_NoLeader(t *testing.T) {
 		{
 			"PUT",
 			"/v1/acl/policy",
+		},
+		{
+			"GET",
+			"/v1/acl/tokens",
 		},
 		{
 			"PUT",
@@ -1496,6 +1503,8 @@ func TestRun_ClientTokensRetry(t *testing.T) {
 			numPolicyCalls++
 		case "/v1/agent/self":
 			fmt.Fprintln(w, `{"Config": {"Datacenter": "dc1", "PrimaryDatacenter": "dc1"}}`)
+		case "/v1/acl/tokens":
+			fmt.Fprintln(w, `[]`)
 		default:
 			fmt.Fprintln(w, "{}")
 		}
@@ -1531,6 +1540,10 @@ func TestRun_ClientTokensRetry(t *testing.T) {
 			"/v1/acl/policy",
 		},
 		{
+			"GET",
+			"/v1/acl/tokens",
+		},
+		{
 			"PUT",
 			"/v1/acl/token",
 		},
@@ -1558,8 +1571,8 @@ func TestRun_ClientTokensRetry(t *testing.T) {
 	}, consulAPICalls)
 }
 
-// Test if there is an old bootstrap Secret we assume the servers were
-// bootstrapped already and continue on to the next step.
+// Test if there is an old bootstrap Secret we still try to create and set
+// server tokens.
 func TestRun_AlreadyBootstrapped(t *testing.T) {
 	t.Parallel()
 	require := require.New(t)
@@ -1581,6 +1594,8 @@ func TestRun_AlreadyBootstrapped(t *testing.T) {
 		switch r.URL.Path {
 		case "/v1/agent/self":
 			fmt.Fprintln(w, `{"Config": {"Datacenter": "dc1", "PrimaryDatacenter": "dc1"}}`)
+		case "/v1/acl/tokens":
+			fmt.Fprintln(w, `[]`)
 		default:
 			// Send an empty JSON response with code 200 to all calls.
 			fmt.Fprintln(w, "{}")
@@ -1629,8 +1644,24 @@ func TestRun_AlreadyBootstrapped(t *testing.T) {
 
 	// Test that the expected API calls were made.
 	require.Equal([]APICall{
-		// We only expect the calls for creating client tokens
-		// and updating the server policy.
+		// We expect calls for updating the server policy, setting server tokens,
+		// and updating client policy.
+		{
+			"PUT",
+			"/v1/acl/policy",
+		},
+		{
+			"GET",
+			"/v1/acl/tokens",
+		},
+		{
+			"PUT",
+			"/v1/acl/token",
+		},
+		{
+			"PUT",
+			"/v1/agent/token/agent",
+		},
 		{
 			"GET",
 			"/v1/agent/self",
@@ -1641,13 +1672,84 @@ func TestRun_AlreadyBootstrapped(t *testing.T) {
 		},
 		{
 			"PUT",
-			"/v1/acl/policy",
-		},
-		{
-			"PUT",
 			"/v1/acl/token",
 		},
 	}, consulAPICalls)
+}
+
+// Test if there is an old bootstrap Secret and the server token exists
+// that we don't try and recreate the token.
+func TestRun_AlreadyBootstrapped_ServerTokenExists(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
+
+	// First set everything up with ACLs bootstrapped.
+	bootToken := "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee"
+	k8s, testAgent := completeBootstrappedSetup(t, bootToken)
+	setUpK8sServiceAccount(t, k8s, ns)
+	defer testAgent.Stop()
+	k8s.CoreV1().Secrets(ns).Create(context.Background(), &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: resourcePrefix + "-bootstrap-acl-token",
+		},
+		Data: map[string][]byte{
+			"token": []byte(bootToken),
+		},
+	}, metav1.CreateOptions{})
+
+	consulClient, err := api.NewClient(&api.Config{
+		Address: testAgent.HTTPAddr,
+		Token:   bootToken,
+	})
+	require.NoError(err)
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI:        ui,
+		clientset: k8s,
+	}
+
+	// Create the server policy and token _before_ we run the command.
+	agentPolicyRules, err := cmd.agentRules()
+	require.NoError(err)
+	policy, _, err := consulClient.ACL().PolicyCreate(&api.ACLPolicy{
+		Name:        "agent-token",
+		Description: "Agent Token Policy",
+		Rules:       agentPolicyRules,
+	}, nil)
+	require.NoError(err)
+	_, _, err = consulClient.ACL().TokenCreate(&api.ACLToken{
+		Description: fmt.Sprintf("Server Token for %s", strings.Split(testAgent.HTTPAddr, ":")[0]),
+		Policies: []*api.ACLTokenPolicyLink{
+			{
+				Name: policy.Name,
+			},
+		},
+	}, nil)
+	require.NoError(err)
+
+	// Run the command.
+	cmdArgs := []string{
+		"-timeout=1m",
+		"-k8s-namespace", ns,
+		"-server-address", strings.Split(testAgent.HTTPAddr, ":")[0],
+		"-server-port", strings.Split(testAgent.HTTPAddr, ":")[1],
+		"-resource-prefix", resourcePrefix,
+	}
+
+	responseCode := cmd.Run(cmdArgs)
+	require.Equal(0, responseCode, ui.ErrorWriter.String())
+
+	// Check that only one server token exists, i.e. it didn't create an
+	// extra token.
+	tokens, _, err := consulClient.ACL().TokenList(nil)
+	require.NoError(err)
+	count := 0
+	for _, token := range tokens {
+		if len(token.Policies) == 1 && token.Policies[0].Name == policy.Name {
+			count++
+		}
+	}
+	require.Equal(1, count)
 }
 
 // Test if there is a provided bootstrap we skip bootstrapping of the servers

--- a/control-plane/subcommand/server-acl-init/servers.go
+++ b/control-plane/subcommand/server-acl-init/servers.go
@@ -5,17 +5,53 @@ import (
 	"fmt"
 	"strings"
 
-	"github.com/hashicorp/consul-k8s/control-plane/consul"
-	"github.com/hashicorp/consul-k8s/control-plane/subcommand/common"
 	"github.com/hashicorp/consul/api"
 	apiv1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/hashicorp/consul-k8s/control-plane/consul"
+	"github.com/hashicorp/consul-k8s/control-plane/subcommand/common"
 )
 
 // bootstrapServers bootstraps ACLs and ensures each server has an ACL token.
-func (c *Command) bootstrapServers(serverAddresses []string, bootTokenSecretName, scheme string) (string, error) {
+// If bootstrapToken is not empty then ACLs are already bootstrapped.
+func (c *Command) bootstrapServers(serverAddresses []string, bootstrapToken, bootTokenSecretName, scheme string) (string, error) {
 	// Pick the first server address to connect to for bootstrapping and set up connection.
 	firstServerAddr := fmt.Sprintf("%s:%d", serverAddresses[0], c.flagServerPort)
+
+	if bootstrapToken == "" {
+		var err error
+		bootstrapToken, err = c.bootstrapACLs(firstServerAddr, scheme, bootTokenSecretName)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// Override our original client with a new one that has the bootstrap token
+	// set.
+	consulClient, err := consul.NewClient(&api.Config{
+		Address: firstServerAddr,
+		Scheme:  scheme,
+		Token:   bootstrapToken,
+		TLSConfig: api.TLSConfig{
+			Address: c.flagConsulTLSServerName,
+			CAFile:  c.flagConsulCACert,
+		},
+	})
+	if err != nil {
+		return "", fmt.Errorf("creating Consul client for address %s: %s", firstServerAddr, err)
+	}
+
+	// Create new tokens for each server and apply them.
+	if err := c.setServerTokens(consulClient, serverAddresses, bootstrapToken, scheme); err != nil {
+		return "", err
+	}
+	return bootstrapToken, nil
+}
+
+// bootstrapACLs makes the ACL bootstrap API call and writes the bootstrap token
+// to a kube secret.
+func (c *Command) bootstrapACLs(firstServerAddr string, scheme string, bootTokenSecretName string) (string, error) {
 	consulClient, err := consul.NewClient(&api.Config{
 		Address: firstServerAddr,
 		Scheme:  scheme,
@@ -29,13 +65,13 @@ func (c *Command) bootstrapServers(serverAddresses []string, bootTokenSecretName
 	}
 
 	// Call bootstrap ACLs API.
-	var bootstrapToken []byte
+	var bootstrapToken string
 	var unrecoverableErr error
 	err = c.untilSucceeds("bootstrapping ACLs - PUT /v1/acl/bootstrap",
 		func() error {
 			bootstrapResp, _, err := consulClient.ACL().Bootstrap()
 			if err == nil {
-				bootstrapToken = []byte(bootstrapResp.SecretID)
+				bootstrapToken = bootstrapResp.SecretID
 				return nil
 			}
 
@@ -69,36 +105,13 @@ func (c *Command) bootstrapServers(serverAddresses []string, bootTokenSecretName
 					Name: bootTokenSecretName,
 				},
 				Data: map[string][]byte{
-					common.ACLTokenSecretKey: bootstrapToken,
+					common.ACLTokenSecretKey: []byte(bootstrapToken),
 				},
 			}
 			_, err := c.clientset.CoreV1().Secrets(c.flagK8sNamespace).Create(c.ctx, secret, metav1.CreateOptions{})
 			return err
 		})
-	if err != nil {
-		return "", err
-	}
-
-	// Override our original client with a new one that has the bootstrap token
-	// set.
-	consulClient, err = consul.NewClient(&api.Config{
-		Address: firstServerAddr,
-		Scheme:  scheme,
-		Token:   string(bootstrapToken),
-		TLSConfig: api.TLSConfig{
-			Address: c.flagConsulTLSServerName,
-			CAFile:  c.flagConsulCACert,
-		},
-	})
-	if err != nil {
-		return "", fmt.Errorf("creating Consul client for address %s: %s", firstServerAddr, err)
-	}
-
-	// Create new tokens for each server and apply them.
-	if err := c.setServerTokens(consulClient, serverAddresses, string(bootstrapToken), scheme); err != nil {
-		return "", err
-	}
-	return string(bootstrapToken), nil
+	return bootstrapToken, err
 }
 
 // setServerTokens creates policies and associated ACL token for each server
@@ -109,9 +122,14 @@ func (c *Command) setServerTokens(consulClient *api.Client, serverAddresses []st
 		return err
 	}
 
+	existingTokens, _, err := consulClient.ACL().TokenList(nil)
+	if err != nil {
+		return err
+	}
+
 	// Create agent token for each server agent.
 	for _, host := range serverAddresses {
-		var token *api.ACLToken
+		var tokenSecretID string
 
 		// We create a new client for each server because we need to call each
 		// server specifically.
@@ -128,26 +146,44 @@ func (c *Command) setServerTokens(consulClient *api.Client, serverAddresses []st
 			return err
 		}
 
-		// Create token for the server
-		err = c.untilSucceeds(fmt.Sprintf("creating server token for %s - PUT /v1/acl/token", host),
-			func() error {
-				tokenReq := api.ACLToken{
-					Description: fmt.Sprintf("Server Token for %s", host),
-					Policies:    []*api.ACLTokenPolicyLink{{Name: agentPolicy.Name}},
+		tokenDescription := fmt.Sprintf("Server Token for %s", host)
+
+		// Check if the token was already created. We're matching on the description
+		// since that's the only part that's unique.
+		for _, t := range existingTokens {
+			if len(t.Policies) == 1 && t.Policies[0].Name == agentPolicy.Name {
+				if t.Description == tokenDescription {
+					tokenSecretID = t.SecretID
+					break
 				}
-				var err error
-				token, _, err = serverClient.ACL().TokenCreate(&tokenReq, nil)
-				return err
-			})
-		if err != nil {
-			return err
+			}
 		}
 
-		// Pass out agent tokens to servers.
-		// Update token.
+		// Create token for the server if it doesn't already exist.
+		if tokenSecretID == "" {
+			err = c.untilSucceeds(fmt.Sprintf("creating server token for %s - PUT /v1/acl/token", host),
+				func() error {
+					tokenReq := api.ACLToken{
+						Description: tokenDescription,
+						Policies:    []*api.ACLTokenPolicyLink{{Name: agentPolicy.Name}},
+					}
+					token, _, err := serverClient.ACL().TokenCreate(&tokenReq, nil)
+					if err != nil {
+						return err
+					}
+					tokenSecretID = token.SecretID
+					return nil
+				})
+			if err != nil {
+				return err
+			}
+		}
+
+		// Pass out agent tokens to servers. It's okay to make this API call
+		// even if the server already has a token since the call is idempotent.
 		err = c.untilSucceeds(fmt.Sprintf("updating server token for %s - PUT /v1/agent/token/agent", host),
 			func() error {
-				_, err := serverClient.Agent().UpdateAgentACLToken(token.SecretID, nil)
+				_, err := serverClient.Agent().UpdateAgentACLToken(tokenSecretID, nil)
 				return err
 			})
 		if err != nil {

